Onboarding modal

ArchitGajjar opened the pull request:
<!-- Thanks for contributing to WordPress Playground! -->

## What?
<!-- In a few words, what is the PR actually doing? Include screenshots or screencasts if applicable -->
* Trying to make client onboarding experience better with this [#497](https://github.com/WordPress/wordpress-playground/issues/497) ticket.
* This code changes will provide a  friendly message(modal experience) to clients that the changes they are making are private and staying within this session. If you refresh your page, all your changes will go away.
* I have introduced new flag called `isOnboarded` inside browser local storage. If this flag is not available into user's browser local storage then it will open modal and if it is present then it will not.

## Why?

<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
* Better client experience and providing a friendly reminder to clients regarding they activities are temporary so they do not hit refresh by mistake and lose their session data.

## How?

<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
* Create custom `React Modal Component` which accepts React Children component with custom message. This modal component has close button inside too so client can easily close the modal after read. 
* Create `util` folder and added utility to check whether the client is onboarded for the first time or not.

## Testing Instructions

<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Check out the branch. -->
<!-- 2. Run a command. -->
<!-- 3. etc. -->
1. Please checkout to my branch `playground/Onboard_enhacement`
2. Install all the dependencies by `npm install`
3. Please run `npm run dev`
4. You will see the modal experienced. This modal is responsive too.
5. Once you close the modal, it will not prompt again if you refresh your browser. 
6. If you still would like to see the modal for testing purposes, please open `dev-tools` and go to `Application` tab and delete the flag "isOnboarded" from `local storage` and then refresh the page. You will be able to see the Modal component again.

## Client Experience 


https://github.com/WordPress/wordpress-playground/assets/5431237/325febb4-7582-4260-a35a-df69ecad9dd8

dmsnell commented:
Would a toast type message or a banner that slides down from the top be more appropriate here? I'm worried about how a modal could interfere with normal use of the app. Even storing a preference in localstorage, people will be running these in private tabs and blank browser sessions.

ArchitGajjar commented:
@dmsnell - Thank you for your input. Have couple of things to ask/share. 
1. Do you have any designs regarding this feature ? where I can look into or any example of web I can refer too. This will reduce communication back and forth.
2. Where would you like to store the preference ?  in the Cookie, or in the URL by query params. Please feel free to share your thoughts further.

dmsnell commented:
Material has its [Toast](https://m2.material.io/components/snackbars) and snackbar, 18F created their [banners](https://designsystem.digital.gov/components/banner/). VSCode uses a toast + snackbar system.

> Where would you like to store the preference ? in the Cookie, or in the URL by query params. Please feel free to share your thoughts further.

it could be better to make this opt-in, and a query arg isn't a bad place to do it. most importantly is finding a way to surface some information like this without taking over the UI or interrupting the typical experience, something a modal or a default-on alert might do

adamziel commented:
I agree with @dmsnell, the original issue said modal but then actually seeing it as a modal doesn't feel right. A query param sounds like a great place to opt-in to that – every app embedding WordPress Playground could decide on their own.

adamziel commented:
Related – https://github.com/WordPress/wordpress-playground/pull/548 explores a modal that opens right away. That one has a different purpose and is also different from what we've just discussed here. Still, I still wanted to surface that one here.

ArchitGajjar commented:
@dmsnell & @adamziel  Thank you for your suggestions and thoughts. I have made the changes and integrated [toast](https://fkhadra.github.io/react-toastify/introduction) with the application. Please review the code and share your feedback. 

## Flow:
1. If query parameter pass `?onBoarding=1` then only the `toast` notification will shows up. 
2.  toast message has default timeout to `20` seconds, it is configurable, I can change it too. 
3. Once the toast message is shown, we will add flag into cookie `onBoarding=1`
4. Evert time, we will refresh, application will check if the cookie value is set `onBoarding` to `1` - if it is set, then we will not show the toast message again. 

## Small announcement: 

-  There was a piece of code written already into the application and was showing the message [here](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/playground/website/src/components/browser-chrome/index.tsx#L44-L48) ,every time we render or re-render the application. As I have implemented the `toast`, I removed that small piece of code which does the same thing I did in this PR to remove duplication.
- Please let me know, if it is okay to clean up, I can revert too.  

## Look & feel

![Screenshot 2023-06-19 at 4 07 18 PM](https://github.com/WordPress/wordpress-playground/assets/11539481/6ac4f7af-4e7f-4e62-b069-db1a859af993)

dmsnell reviewed (commented):
@ArchitGajjar I like the looks of that toast, and I think this is heading in the right direction.

 - the docblock comments are helpful, but we don't generally need the `@description` line. these can be removed. also we use the `@param` tag instead of the `argument` tag, and no colons `:` after the tags. you can review any of the existing docblocks for examples.
 - we don't need to document _everything_, especially if it's already documented elsewhere or if the documentation doesn't add information that's already in the code. it's there for what _isn't_ there, not for what is. this means that `@returns: None` aren't needed (and others)
 - I'm concerned about the use of storing the preference in a cookie. we also have `localstorage` available to us that's much simpler to read from and write to, and cookies are sent back and forth with the server.
 - did you consider moving the entire logic of the onboarding modal into its own component instead of putting some logic in browser-chrome, some in the toast component, and some in the utils? the Toast component is nice to have as a generic thing, but instead of `showToast && <Toast />` we could also have `<OnboardingModal />` in this location that unconditionally renders, but whose render function determines whether or not to render
 - there's no need to have `Toast/Toast.tsx` and an empty `index.tsx` - let's put the toast logic into `index.tsx` until and unless it gets more complicated and there's a clear need to split them.

ArchitGajjar commented:
Thank you @dmsnell for reviewing, I will make those changes and ask you for re-review again.

ArchitGajjar commented:
@dmsnell & @adamziel  Thank you for your help in review. could you please take a look once, I have addressed your comments, please let me know if I missed anything or I need to fix in my latest changes.

ArchitGajjar commented:
Hi @dmsnell & @adamziel - can you please take a look into this and share your feedback. Thank you!

dmsnell commented:
@ArchitGajjar we are both on vacation at the moment but I will attend to this within the next two weeks. thank you!

dmsnell commented on packages/playground/website/src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,44 @@
+	return (
+		<div>{showToast && <ToastContainer className="toast-container" />}</div>
+	);
+};
```
why is the `div` required here?

I guess `<ToastContainer>` is the portal into which toasts render? In that case, maybe it ought to unconditionally render and some new `OnboardingMessage` should render into it?

ArchitGajjar replied:
correct, we do not need extra `<div>` element here. 👍

dmsnell commented on packages/playground/website/src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,44 @@
+	const showToastMessage = (message: string): void => {
+		/**
+		 * @param message: string, options: object
+		 * @see https://fkhadra.github.io/react-toastify/introduction/
```
given that there's a link back to the project through `npm` itself I don't see a need to include this link here, nice as it is. we don't want to end up putting already-discoverable links everywhere we call functions from dependencies otherwise it becomes a bunch of noise in the code that we have to maintain and sort through

ArchitGajjar replied:
Good point, I will remove!

dmsnell commented on packages/playground/website/src/components/Toast/css/toast.css:
```diff
@@ -0,0 +1,13 @@
+	/* The Toast container will be in full screen width */
+	.toast-container {
+		width: 100%;
+		padding-bottom: 60px;
```
this rule looks redundant, since it's already in the `.toast-container`

dmsnell commented on packages/playground/website/src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,44 @@
+	);
+};
+
+export default Toast;
```
I like where you're going with this, but I think I overcommunicated how much I thought belonged in the separate component.

the Toast is unto itself a generic UI component. the onboarding message is a good candidate for something that _uses_ the toast system. image splitting these into separate components.

```jsx
// browser-chrome.tsx
…
<ToastContainer className="toast-container" />
```

```jsx
// onboarding-message.tsx
…

export const OnboardingMessage = () => {
	useEffect( () => {
		…

		if ( showOnboarding ) {
			toast.info( … );
		}
	}, [] );
}
```

We don't need to call `useState` to set the toast message, as it's inherently an imperative API. we can call that once in a `useEffect` with an empty array as its dependencies. Even though the query string in the URL _could_ change, I don't see a need to pop up the onboarding message if it were to dynamically add the expected parameter via JS.

dmsnell commented on packages/playground/website/src/hooks/useQuery.ts:
```diff
@@ -0,0 +1,21 @@
+  return null 
+};
+
+export default useQuery;
```
for now I think it would be best not to add this. for one, it's not a hook in the normal sense, and for two, the app is full of uses of `URLSearchParams` that I would prefer we follow the pattern of here.

if we want to abstract reading query args we can do that in a separate isolated PR.

ArchitGajjar replied:
okay, sure, I will remove it for now..

dmsnell reviewed (commented):
@ArchitGajjar it'd be great to work on the abstraction here a bit, as noted in some comments. if you need help making sense of what I wrote please ask 😄

dmsnell commented on packages/playground/website/src/components/Toast/index.tsx:
```diff
@@ -0,0 +1,44 @@
+		 */
+		toast.info(message, {
+			autoClose: 20000,
+			position: toast.POSITION.BOTTOM_CENTER,
```
let's add `hideProgressBar` here - it's a bit distracting to see it move around

ArchitGajjar replied:
sure, will add here this prop. Thank you!

dmsnell reviewed (changes requested):
@ArchitGajjar I also realized that due to the way the Playground now supports persistence we should make sure this only displays when running in "Temporary" mode. if the Playground loads in Persistent or as a Live Directory the warning shouldn't apply or store anything in `localStorage`

ArchitGajjar commented:
Thank you @dmsnell  for reviewing through, I will review all the comments and update here!

adamziel commented:
This PR seems stale, I'll close it for now. Feel free to reopen.

ArchitGajjar commented:
I am back to work on this issue, apologies got delayed due to some health issue. 

I will re-open this PR soon. Thank you for your patience and support!

dmsnell commented:
@ArchitGajjar hope you are better. no need for apologies :)
